tiledb/common/evaluator/test/unit_evaluator.cc

```diff
@@ -0,0 +1,389 @@
+
+  first_thread = false;
+  std::thread t2([&eval]() {
+    // This should wait within the evaluator until the value becomes available
+    eval("key");
+  });
+  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+
+  // Check a single entry is in progress and no ready entries are present
+  CHECK(eval.policy().entries_size() == 0);
+  CHECK(eval.policy().inprogress_size() == 1);
+  CHECK(eval.policy().entry_in_progress("key"));
+
+  // Unblock the first thread so that the evaluation can finish and f(key) gets
+  // cached, second reader gets the value as well and becomes unblocked
+  finish_I_promise.set_value();
+
+  t.join();
+  t2.join();
+
+  // Make sure the second reader got a cached value and didn't just execute the
+  // callback again.
+  CHECK(second_exec == false);
+}
+
+TEST_CASE("Memory budgeting policy is enforced", "[evaluator][memory_budget]") {
+  size_t budget = 4096;
+  auto f = [](const std::string& key) { return key; };
+  auto sizefn = [](const std::string& val) {
+    return sizeof(std::string) + val.size();
+  };
+
+  WhiteBoxEvaluator<
+      WhiteBoxMemoryBudgetPolicy<std::string, std::string>,
+      decltype(f)>
+      mem_budgeted_eval(f, sizefn, budget);
+
+  CHECK(mem_budgeted_eval.policy().memory_consumed() == 0);
+
+  size_t num_entries = 0;
+  size_t c = 0;
+  std::string str = "key0";
+  size_t unit_size = sizeof(std::shared_ptr<std::string>) + sizefn(str);
+  while (c < budget - unit_size) {
+    str = "key" + std::to_string(c);
+    unit_size = sizeof(std::shared_ptr<std::string>) + sizefn(str);
+
+    mem_budgeted_eval(str);
+    CHECK(mem_budgeted_eval.policy().memory_consumed() == c + unit_size);
+
+    ++num_entries;
+    c += unit_size;
+  }
+
+  // No eviction happened
+  CHECK(mem_budgeted_eval.policy().entries_size() == num_entries);
+
+  // str should be over budget, eviction should happen
+  str = "key" + std::to_string(c);
+  unit_size = sizeof(std::shared_ptr<std::string>) + sizefn(str);
+  std::string evicted_val = "key0";
+  size_t evicted_size =
+      sizeof(std::shared_ptr<std::string>) + sizefn(evicted_val);
+  mem_budgeted_eval(str);
+  CHECK(mem_budgeted_eval.policy().entries_size() == num_entries);
+  CHECK(
+      mem_budgeted_eval.policy().memory_consumed() ==
+      c + unit_size - evicted_size);
+
+  // The policy evicts multiple entries until the new value fits
+  str = std::string(unit_size, 'a');
+  mem_budgeted_eval(str);  // evicts 2 lru entries
+  CHECK(mem_budgeted_eval.policy().entries_size() == num_entries - 1);
+  CHECK(mem_budgeted_eval.policy().memory_consumed() <= budget);
+}
+
+TEST_CASE("Test various corner cases", "[evaluator][basic]") {
+  size_t budget = 0;
+  auto f = [](const std::string& key) { return key; };
+  auto sizefn = [](const std::string& val) {
+    return sizeof(std::string) + val.size();
+  };
+
+  auto matcher = Catch::Matchers::ContainsSubstring("greater than zero");
+  REQUIRE_THROWS_WITH(
```

Comment:
I consider it overkill and slightly bad to validate the text of an exception. What really matters is that it throws---that's the black box test.

A white box test might want to distinguish one `throw` statement from another, in which case something like that is warranted. That doesn't seem to be the case here.